src/main/java/org/sonar/plugins/jacoco/WildcardPatternFileScanner.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * SonarQube JaCoCo Plugin
+ * Copyright (C) 2018-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.jacoco;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+import org.sonar.api.internal.google.common.annotations.VisibleForTesting;
+import org.sonar.api.utils.WildcardPattern;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+public final class WildcardPatternFileScanner {
+
+  private static final Logger LOG = Loggers.get(WildcardPatternFileScanner.class);
+
+  private static final int SEARCH_MAX_DEPTH = 64;
+
+  private static final String PATH_MATCHER_SPECIAL_CHAR = "*?";
+
+  private WildcardPatternFileScanner() {
+    // utility class
+  }
+
+  public static List<Path>  scan(Path baseDirectory, String patternPath) {
+    String unixLikePatternPath = toUnixLikePath(patternPath);
+    int specialCharIndex = indexOfMatcherSpecialChar(unixLikePatternPath);
+    if (specialCharIndex == -1) {
+      return scanNonWildcardPattern(baseDirectory, unixLikePatternPath);
+    } else {
+      int additionalBaseDirectoryPart = unixLikePatternPath.lastIndexOf('/', specialCharIndex);
+      if (additionalBaseDirectoryPart != -1) {
+        Path additionalBaseDirectory = toFileSystemPath(unixLikePatternPath.substring(0, additionalBaseDirectoryPart + 1));
+        String remainingWildcardPart = unixLikePatternPath.substring(additionalBaseDirectoryPart + 1);
+        Path moreSpecificBaseDirectory = baseDirectory.resolve(additionalBaseDirectory);
+        return scanWildcardPattern(moreSpecificBaseDirectory, remainingWildcardPart);
+      } else {
+        return scanWildcardPattern(baseDirectory, unixLikePatternPath);
+      }
```

Comment:
Basically, this code is an optimization to avoid walking through every file from the baseDir, but to start at the first wildcard, correct?

I believe this deserves a comment briefly explaining this.